crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -147,24 +551,42 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
 /// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
 /// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
 pub fn apply_state_transition<S: States>(world: &mut World) {
-    // We want to take the `NextState` resource,
-    // but only mark it as changed if it wasn't empty.
-    let mut next_state_resource = world.resource_mut::<NextState<S>>();
-    if let Some(entered) = next_state_resource.bypass_change_detection().0.take() {
-        next_state_resource.set_changed();
-
-        let mut state_resource = world.resource_mut::<State<S>>();
-        if *state_resource != entered {
+    let Some(next_state_resource) = world.get_resource::<NextState<S>>() else {
+        return;
+    };
+    let Some(current_state) = world.get_resource::<State<S>>().map(|s| s.0.clone()) else {
+        return;
+    };
+    let entered = match next_state_resource {
+        NextState::Keep => None,
+        NextState::Value(v) => Some(v.clone()),
+        NextState::Setter(f) => Some(f(current_state.clone())),
```

Comment:
I think our function `f` should take the current state by reference. The ergonomics shouldn't be worse, and I can't imagine why you'd want full ownership here.

Reply:
The reason was to add clarity to the idea that this function converts from one state to another. I don't think there will be much of a difference in ergonomics (other than if you wind up with a no-op), but passing in an owned value felt like it communicates "take the current state, and give us the new one" better.

Reply:
Sure :) No strong feelings here.

Reply:
I will see if there's a way to pass in the actual current value rather than cloning, but otherwise considering this resolved :)